cou/apps/app.py

```diff
@@ -182,7 +305,251 @@ def _get_os_origin(self) -> str:
+    def post_upgrade_plan(self, target: OpenStackRelease) -> list[UpgradeStep]:
+        """Post Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: Plan that will add post upgrade as sub steps.
+        :rtype: list[UpgradeStep]
+        """
+        return [self._get_reached_expected_target_plan(target)]
+
+    def generate_upgrade_plan(self, target: str) -> UpgradeStep:
+        """Generate full upgrade plan for an Application.
+
+        :param target: OpenStack codename to upgrade.
+        :type target: str
+        :return: Full upgrade plan if the Application is able to generate it.
+        :rtype: UpgradeStep
+        """
+        target_version = OpenStackRelease(target)
+        upgrade_plan = UpgradeStep(
+            description=(
+                f"Upgrade plan for '{self.name}' from: {self.current_os_release} " f"to {target}"
+            ),
+            parallel=False,
+            function=None,
+        )
+        all_steps = (
+            self.pre_upgrade_plan(target_version)
+            + self.upgrade_plan(target_version)
+            + self.post_upgrade_plan(target_version)
+        )
+        for step in all_steps:
+            if step:
+                upgrade_plan.add_step(step)
+        return upgrade_plan
+
+    def _get_refresh_charm_plan(
+        self, target: OpenStackRelease, parallel: bool = False
+    ) -> Optional[UpgradeStep]:
+        """Get plan for refreshing the current channel.
+
+        This function also identifies if charm comes from charmstore and in that case,
+        makes the migration.
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param parallel: Parallel running, defaults to False
+        :type parallel: bool, optional
+        :return: Plan for refreshing the charm.
+        :rtype: Optional[UpgradeStep]
+        """
+        switch = None
+        description = (
+            f"Changing '{self.name}' channel from: '{self.channel}' "
+            f"to: '{self.expected_current_channel}'"
+        )
+
+        try:
+            # get the OpenStack release from the channel track of the application.
+            os_track_release_channel = OpenStackRelease(self.channel.split("/", maxsplit=1)[0])
+        except ValueError:
+            logger.debug("Current channel it's inexistent or it's not on expected format")
+            os_track_release_channel = self.current_os_release
+
+        if self.charm_origin == "cs":
+            description = f"Migration of '{self.name}' from charmstore to charmhub"
+            switch = f"ch:{self.charm}"
+        elif self.channel == self.expected_current_channel:
+            description = (
+                f"Refresh '{self.name}' to the latest revision of "
+                f"'{self.expected_current_channel}'"
+            )
+        elif os_track_release_channel >= target:
+            logger.warning(
```

Comment:
this may not be a warning, especially if it's the result of rerunning cou after a partial upgrade. How about `info`?